### PR TITLE
Support masked sum on CSR tensors [CPU, CUDA]

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75506
* __->__ #72633

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 1]
```diff
@@ -94,6 +94,9 @@
 #include <ATen/ops/zeros.h>
 #endif
 
+#include <ATen/ops/_unique.h>
+#include <ATen/ops/tensor.h>
```
Comment (peterbell10) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:98]: FYI: These need to come before the `#endif` above. That's why the `no-ops` CI is failing.

Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp]: Fixed. Thanks, @peterbell10 !

File: torch/_masked/__init__.py [hunk 2]
```diff
@@ -697,6 +704,9 @@ def sum(input: Tensor,
             result = torch.sparse_coo_tensor(indices, result._values(), shape, dtype=result.dtype, device=result.device)
 
         return result
+
+    elif input.layout == torch.sparse_csr:
+        return torch._sparse_csr_sum(mask_input, dim=list(dim_), keepdim=bool(keepdim), dtype=dtype)
     else:
         raise ValueError(f'masked sum expects strided or sparse_coo tensor (got {input.layout} tensor)')
```
Comment (cpuhrsch) [inline torch/_masked/__init__.py:711]: nit: I guess this could be updated

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -1003,6 +1003,13 @@ def _generate_reduction_kwargs(ndim, supports_multiple_dims=True):
     # Test reducing inner and outer most dimensions
     yield {'dim': 0, 'keepdim': True}
     yield {'dim': -1, 'keepdim': False}
+    # Reductions of CSR tensors use different implementations for
+    # inner and/or outer dimensions. So, as a minimum of testing CSR
+    # implementations the following kwargs must be generated:
+    #   dict(dim=0, keepdim=True)
+    #   dict(dim=1, keepdim=True)
+    #   dict(dim=(0, 1), keepdim=True)
+    yield {'dim': -1, 'keepdim': True}
```
Comment (cpuhrsch) [inline torch/testing/_internal/common_methods_invocations.py:1012]: This make sense, but it's annoying that *all* reduction tests will be updated. Make we can constrain this to `sample_inputs_masked_reduction` only for now?

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 4]
```diff
@@ -685,5 +687,341 @@ TORCH_IMPL_FUNC(_convert_indices_from_csr_to_coo_structured_cpu) (
...
+  if (dims.size() == 0) {
+    // after gh-29137 is resolved, delete this if-block
+    dims.emplace_back(0);
+    dims.emplace_back(1);
+  }
+  return reduce_sparse_csr_cpu_template<scalar_t>(sparse, dims, rop);
+}
+
+template <typename scalar_t>
+struct ReductionAddOp {
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:1005]: Looks like you're already well setup for additional reductions

**Review by cpuhrsch — APPROVED** [review]
Generally this looks good. I didn't do a close read for performance of the CUDA kernel. See some more comments to address that I added. 

Overall this work should be immediately unifiable with [regular structured torch.sum for CSR](https://github.com/pytorch/pytorch/pull/74766) and [segment reduce](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SegmentReduce.cpp).

**Comment by cpuhrsch** [comment]
@pearu - is this ready? I had added a few comments.

File: torch/_masked/__init__.py (hunk 2 above)
Comment (pearu) [reply torch/_masked/__init__.py:711]: Fixed.

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (pearu) [reply torch/testing/_internal/common_methods_invocations.py]: The constraint is now implemented within `sample_inputs_sparse_csr_masked_reduction`.

**Comment by janeyx99** [comment]
@pytorchbot revert this

Sorry, reverting because this seemed to introduce a failing test in our slow_test config https://hud.pytorch.org/minihud?name_filter=linux-bionic-py3.7-clang9-slow%20/%20test%20(slow,%201,%201,%20linux.2xlarge)#cda3f586d0b1189c5c4175a503f99f374d1fcce2.

Please add the ciflow/trunk label when you reland this PR.

**Comment by cpuhrsch** [comment]
Reopening it to attempt reland

**Comment by pearu** [comment]
@cpuhrsch could you try landing this now.

**Comment by b0noI** [comment]
@pytorchbot revert this . still no, errors:

Summary: 
stderr: caffe2/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:1041:8: error: unused variable 'ncols' [-Werror,-Wunused-variable]
  auto ncols = sparse.size(1);
       ^
caffe2/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:1135:8: error: unused variable 'numel' [-Werror,-Wunused-variable]
  auto numel = values.numel();
       ^
stderr: caffe2/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:1041:8: error: unused variable 'ncols' [-Werror,-Wunused-variable]
  auto ncols = sparse.size(1);
       ^
caffe2/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:1135:8: error: unused variable 'numel' [-Werror,-Wunused-variable]
  auto numel = values.numel();
       ^

**Comment by b0noI** [comment]
please double check all unused variables

**Comment by malfet** [comment]
Hmm, somehow `-Wno-unused-variable` were added to cmake build config:
```
/usr/bin/clang++ -DADD_BREAKPAD_SIGNAL_HANDLER -DAT_PER_OPERATOR_HEADERS -DCPUINFO_SUPPORTED_PLATFORM=1 -DFMT_HEADER_ONLY=1 -DFXDIV_USE_INLINE_ASSEMBLY=0 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DNNP_CONVOLUTION_ONLY=0 -DNNP_INFERENCE_ONLY=0 -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DUSE_C10D_GLOO -DUSE_C10D_MPI -DUSE_DISTRIBUTED -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -Dtorch_cpu_EXPORTS -I/fsx/users/nshulga/pytorch/build/aten/src -I/fsx/users/nshulga/pytorch/aten/src -I/fsx/users/nshulga/pytorch/build -I/fsx/users/nshulga/pytorch -I/fsx/users/nshulga/pytorch/cmake/../third_party/benchmark/include -I/fsx/users/nshulga/pytorch/cmake/../third_party/cudnn_frontend/include -I/fsx/users/nshulga/pytorch/third_party/onnx -I/fsx/users/nshulga/pytorch/build/third_party/onnx -I/fsx/users/nshulga/pytorch/third_party/foxi -I/fsx/users/nshulga/pytorch/build/third_party/foxi -I/fsx/users/nshulga/pytorch/torch/csrc/api -I/fsx/users/nshulga/pytorch/torch/csrc/api/include -I/fsx/users/nshulga/pytorch/caffe2/aten/src/TH -I/fsx/users/nshulga/pytorch/build/caffe2/aten/src/TH -I/fsx/users/nshulga/pytorch/build/caffe2/aten/src -I/fsx/users/nshulga/pytorch/caffe2/../third_party -I/fsx/users/nshulga/pytorch/caffe2/../third_party/breakpad/src -I/fsx/users/nshulga/pytorch/build/caffe2/../aten/src -I/fsx/users/nshulga/pytorch/torch/csrc -I/fsx/users/nshulga/pytorch/third_party/miniz-2.0.8 -I/fsx/users/nshulga/pytorch/third_party/kineto/libkineto/include -I/fsx/users/nshulga/pytorch/third_party/kineto/libkineto/src -I/fsx/users/nshulga/pytorch/torch/csrc/distributed -I/fsx/users/nshulga/pytorch/aten/../third_party/catch/single_include -I/fsx/users/nshulga/pytorch/aten/src/ATen/.. -I/fsx/users/nshulga/pytorch/third_party/FXdiv/include -I/fsx/users/nshulga/pytorch/c10/.. -I/fsx/users/nshulga/pytorch/third_party/pthreadpool/include -I/fsx/users/nshulga/pytorch/third_party/cpuinfo/include -I/fsx/users/nshulga/pytorch/third_party/QNNPACK/include -I/fsx/users/nshulga/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/include -I/fsx/users/nshulga/pytorch/aten/src/ATen/native/quantized/cpu/qnnpack/src -I/fsx/users/nshulga/pytorch/third_party/cpuinfo/deps/clog/include -I/fsx/users/nshulga/pytorch/third_party/NNPACK/include -I/fsx/users/nshulga/pytorch/third_party/fbgemm/include -I/fsx/users/nshulga/pytorch/third_party/fbgemm -I/fsx/users/nshulga/pytorch/third_party/fbgemm/third_party/asmjit/src -I/fsx/users/nshulga/pytorch/third_party/FP16/include -I/fsx/users/nshulga/pytorch/third_party/tensorpipe -I/fsx/users/nshulga/pytorch/build/third_party/tensorpipe -I/fsx/users/nshulga/pytorch/third_party/tensorpipe/third_party/libnop/include -I/fsx/users/nshulga/pytorch/third_party/fmt/include -I/fsx/users/nshulga/pytorch/third_party/flatbuffers/include -isystem /fsx/users/nshulga/pytorch/build/third_party/gloo -isystem /fsx/users/nshulga/pytorch/cmake/../third_party/gloo -isystem /fsx/users/nshulga/pytorch/cmake/../third_party/googletest/googlemock/include -isystem /fsx/users/nshulga/pytorch/cmake/../third_party/googletest/googletest/include -isystem /fsx/users/nshulga/pytorch/third_party/protobuf/src -isystem /fsx/users/nshulga/conda/include -isystem /fsx/users/nshulga/pytorch/third_party/gemmlowp -isystem /fsx/users/nshulga/pytorch/third_party/neon2sse -isystem /fsx/users/nshulga/pytorch/third_party/XNNPACK/include -isystem /fsx/users/nshulga/pytorch/cmake/../third_party/eigen -isystem /opt/amazon/openmpi/include -isystem /usr/local/cuda-11.1/include -isystem /fsx/users/nshulga/pytorch/build/include  -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-range-loop-analysis -Wno-pass-failed -Wsign-compare -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-typedef-redefinition -Wno-unknown-warning-option -Wno-unused-private-field -Wno-inconsistent-missing-override -Wno-aligned-allocation-unavailable -Wno-c++14-extensions -Wno-constexpr-not-const -Wno-missing-braces -Qunused-arguments -fcolor-diagnostics -Werror -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -DHAVE_AVX512_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3 -DNDEBUG -DNDEBUG -fPIC -DCAFFE2_USE_GLOO -DTH_HAVE_THREAD -Werror -Wall -Wextra -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-type-limits -Wno-array-bounds -Wno-sign-compare -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-missing-braces -Wno-range-loop-analysis -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -pthread -DASMJIT_STATIC -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp.o -c /fsx/users/nshulga/pytorch/aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp
```
Investigating how it got there
